distributed/tests/test_worker_state_machine.py

```diff
@@ -285,20 +300,163 @@ async def test_fetch_to_compute(c, s, a, b):
+        5. when x finally reaches the top of the data_needed heap, the w1 will not try
+           contacting w2
+
+    as_deps=False
+        1. scheduler calls handle_acquire_replicas(who_has={"x": [w2, w3]}) on w1
+        2. x transitions released -> fetch
+        3. the network stack is busy, so x does not transition to flight yet.
+        4. scheduler calls handle_acquire_replicas(who_has={"x": [w3]}) on w1
+        5. when x finally reaches the top of the data_needed heap, the w1 will not try
+           contacting w2
+    """
+    x = (await c.scatter({"x": 1}, workers=[w2.address, w3.address], broadcast=True))[
+        "x"
+    ]
+    with freeze_data_fetching(w1):
+        if as_deps:
+            y1 = c.submit(inc, x, key="y1", workers=[w1.address])
+        else:
+            s.request_acquire_replicas(w1.address, ["x"], stimulus_id="test")
 
-    await f1
+        await wait_for_state("x", "fetch", w1)
+        assert w1.tasks["x"].who_has == {w2.address, w3.address}
+
+        assert len(s.tasks["x"].who_has) == 2
+        s.handle_missing_data(
+            key="x", worker="na", errant_worker=w2.address, stimulus_id="test"
+        )
+        assert len(s.tasks["x"].who_has) == 1
+
+        if as_deps:
+            y2 = c.submit(inc, x, key="y2", workers=[w1.address])
+        else:
+            s.request_acquire_replicas(w1.address, ["x"], stimulus_id="test")
+
+        while w1.tasks["x"].who_has != {w3.address}:
+            await asyncio.sleep(0.01)
+
+    await wait_for_state("x", "memory", w1)
+
+    assert_story(
+        w1.story("request-dep"),
+        [("request-dep", w3.address, {"x"})],
+        # This tests that there has been no attempt to contact w2.
+        # If the assumption being tested breaks, this will fail 50% of the times.
+        strict=True,
+    )
+
+
+@gen_cluster(client=True, nthreads=[("", 1)] * 2)
+async def test_fetch_to_missing(c, s, a, b):
+    """
+    1. task x is a dependency of y
+    2. scheduler calls handle_compute("y", who_has={"x": [b]}) on a
+    3. x transitions released -> fetch -> flight; a connects to b
+    4. b responds it's busy. x transitions flight -> fetch
+    5. The busy state triggers an RPC call to Scheduler.who_has
+    6. the scheduler responds {"x": []}, because w1 in the meantime has lost the key.
+    7. x is transitioned fetch -> missing
+    """
+    x = (await c.scatter({"x": 1}, workers=[b.address]))["x"]
+    b.total_in_connections = 0
+    with freeze_data_fetching(a):
+        y = c.submit(inc, x, key="y", workers=[a.address])
+        await wait_for_state("x", "fetch", a)
+        # Do not use handle_missing_data, since it would cause the scheduler to call
+        # handle_free_keys(["y"]) on a
+        s.remove_replica(ts=s.tasks["x"], ws=s.workers[b.address])
+        # We used a scheduler internal call, thus corrupting its state.
+        # Don't crash at the end of the test.
+        s.validate = False
```

Comment:
This is an anti pattern and we should not intentionally put the scheduler in a corrupt state. 

Is the condition this test is provoking even real? I would argue that it is not even intended for the key to get into a missing state here but the scheduler should rather ensure the task is released

Reply:
It's real. I rewrote the test.